File: Recursion/leaf-similar-trees.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int>tree_last1;
    vector<int>tree_last2;


    void dfs(TreeNode* root,vector<int> &s)
    {
        if(root->left == NULL && root->right == NULL)
        {
            s.push_back(root->val);
            return;
        }
        else if(root == NULL)
        {
            return;
        }
        else if(root->left == NULL && root->right != NULL)
        {
            
            dfs(root->right,s);
        }
        else if(root->left != NULL && root->right == NULL)
        {
            
            dfs(root->left,s);
        }
        else
        {
            dfs(root->left,s);
            dfs(root->right,s);

        }

    }

    
    bool leafSimilar(TreeNode* root1, TreeNode* root2) {
        
        dfs(root1,tree_last1);
        dfs(root2,tree_last2);

        if(tree_last1.size() != tree_last2.size())
        {
            return false;
        }

        for(int i=0;i<tree_last1.size();i++)
        {
            if(tree_last1[i] != tree_last2[i])
            {
                return false;
            }
        }
        return true;







    }
};

File: Basic-DSA-Quest/leaf-similar-trees.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def leafSimilar(self, root1: Optional[TreeNode], root2: Optional[TreeNode]) -> bool:
        lvs1 = []
        lvs2 = []
        def get_leaves(root,ls):
            if not(root.left) and not(root.right):
                ls.append(root.val)
                return
            elif not(root.left) and root.right:
                get_leaves(root.right,ls)
            elif not(root.right) and root.left:
                get_leaves(root.left,ls)
            else:
                get_leaves(root.left,ls)
                get_leaves(root.right,ls)
        get_leaves(root1,lvs1)
        get_leaves(root2,lvs2)
        return lvs1 == lvs2
    
            
